json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java

```diff
@@ -40,17 +41,12 @@
         if (String.class.equals(type)) {
             return Optional.empty();
         }
-
-        return Optional.of((pos, stmt, ctx) -> {
```

Comment:
I believe this change makes the json mapper rely on the String mapper present at the time of registering the json mapper, and ignore any string mapper registered afterward. Not a huge problem I think but maybe users need to be aware registration order will matter between those two?

Reply:
`ArgumentFactory` is only invoked once we are preparing the statement to execute - you have the (static) argument type at this point, this is far after registration time.

Reply:
Right, I'm stupid, sorry

Reply:
I think it's reasonable to expect the `@Json String` argument factory and `JsonMapper` to be configured prior to executing a statement that depends on it.

Reply:
@stevenschlansker That's not accurate. The `ArgumentFactory.build()` method is invoked by `Arguments.findFor()` which is invoked at the specific chaining method where you bind the value, e.g.

```
handle.createUpdate(sql)
    .bindByType("foo", foo, QualifiedType.of(Foo.class).with(Json.class)) // argument factory invoked here
    .execute();
```

So it's conceivable the if somebody configured the `JsonMapper` or registered the `@Json String` mapper later in the call chain, they wouldn't be used:

```
handle.createUpdate(sql)
    .bindByType("foo", foo, QualifiedType.of(Foo.class).with(Json.class))
    .configure(JsonConfig.class, jc -> jc.setJsonMapper(myMapper)) // too late!
    .execute();
```

I just don't think it's a very likely case, and not worth worrying about unless we hear a bug report on it.

I suspect the vast majority of folks using JSON mapping will just follow the cookbook in the developer guide and configure everything on the `Jdbi` instance.

Reply:
Ah, sorry, I misspoke.  Either way my overall point was that you are already logically building the statement, even if we are not actually building the `PreparedStatement`.